### PR TITLE
[runtime] Add a dumper/verifier for existential containers (i.e. Any)…

### PR DESCRIPTION
**Pull request by gottesmm**
… in c++.

The dumper method dumps:

1. The container's metadata pointer.
2. The container's metadata representation.
3. A pointer to the container's value.
4. Whether or not said value is stored inline in the container.
5. The names of the conformance, protocol of all witness tables that trail the container.

This provides a general overview that can be used even when working with SIL
code in the debugger by grabbing a pointer to swift Anys and then calling the
c++ any method upon them.

The verifier is intended to be used in conjunction with ASAN for maximum
effect to catch use-after-frees of existential boxes. Specifically:

1. The memory where the box itself is stored is tested to see if the memory is
live by performing a load from the box to get the Type field.
2. If the value is stored out of line, the first byte of the value is read
from. Volatile is used to make sure that the optimizer does not remove the
read.

Additionally it goes through all of the protocol conformance descriptor and
runs their respective verification routine as well.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@aschwaighofer Can you take a look at this? Just crossing my Ts.

**Review by aschwaighofer — CHANGES_REQUESTED** [review]

File: stdlib/public/runtime/Metadata.cpp [hunk 1]
```diff
@@ -2474,6 +2493,89 @@ ExistentialCacheEntry::ExistentialCacheEntry(Key key) {
...
+         getNameForExistentialTypeRepresentation(existentialMetadata));
+  printf("Value Pointer: %p.\n",
+         existentialMetadata->projectValue(
+             reinterpret_cast<const OpaqueValue *>(this)));
+
+  bool isValueInline = existentialMetadata->isValueInline(
+      reinterpret_cast<const OpaqueValue *>(this));
+  printf("Is Value Stored Inline: %s.\n", isValueInline ? "true" : "false");
+
+  unsigned numWitnessTables = existentialMetadata->Flags.getNumWitnessTables();
```
[the remark this answers is not in the document]
Comment (aschwaighofer) [reply stdlib/public/runtime/Metadata.cpp]: Apparently, I made the relevant comment further down. We don't have a pointer to the type of the existential. So we can't do this here in a generic way.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@aschwaighofer How does this look?

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Review by aschwaighofer — APPROVED** [review]
Looks good

**Pull request merged**